presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,550 @@
+                        positionsToSkip += 1 + nextPosition - streamPosition;
+                        streamPosition = nextPosition + 1;
+                    }
+                    skipData(streamPosition, positionsToSkip);
+                }
+            }
+        }
+        if (dataToSkip > 0) {
+            dataStream.skip(dataToSkip);
+        }
+        return streamPosition;
+    }
+
+    private int readAllNulls(int[] positions, int positionCount)
+    {
+        if (nonDeterministicFilter) {
+            outputPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                if (filter.testNull()) {
+                    outputPositionCount++;
+                }
+                else {
+                    outputPositionCount -= filter.getPrecedingPositionsToFail();
+                    i += filter.getSucceedingPositionsToFail();
+                }
+            }
+        }
+        else if (nullsAllowed) {
+            outputPositionCount = positionCount;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        allNulls = true;
+        return positions[positionCount - 1] + 1;
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (dataStream == null) {
+            presentStream.skip(items);
+        }
+        else if (presentStream != null) {
+            int lengthToSkip = presentStream.countBitsSet(items);
+            dataStream.skip(lengthStream.sum(lengthToSkip));
+        }
+        else {
+            dataStream.skip(lengthStream.sum(items));
+        }
+    }
+
+    /**
+     * skip the dataStream from the streamPosition to the current position
+     * increments the lengthIndex and
+     * calculate the length sum to the skipped.
+     */
+    private void skipData(int start, int items)
+            throws IOException
+    {
+        int dataToSkip = 0;
+        for (int i = 0; i < items; i++) {
+            if (isNullVector == null || !isNullVector[start + i]) {
+                dataToSkip += lengths[lengthIndex];
+                lengthIndex++;
+            }
+        }
+        dataStream.skip(dataToSkip);
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
```

Comment:
you are introducing new pattern where compactValues used in both getBlock and getBlockView; I like this because it makes the code more succinct; however, it doesn't seem to make sense to create copy of nulls, but not of offsets and not of data; hence, I'd remove nulls parameter from compactValues and refactor this method like so:

```
        if (positionCount != outputPositionCount) {
            compactValues(positions, positionCount, includeNulls);
        }
        
        Block block = new VariableWidthBlock(positionCount, dataAsSlice, offsets, Optional.ofNullable(includeNulls ? nulls : null));
        dataAsSlice = null;
        data = null;
        offsets = null;
        nulls = null;
        return block;
```